driver-core/src/main/java/com/datastax/driver/core/querybuilder/BuiltStatement.java

```diff
@@ -297,14 +296,16 @@ public Boolean isIdempotent() {
     @Override
     public String toString() {
         try {
-            if (forceNoValues) {
-                return getQueryString();
-            }
-            StringBuilder queryString = buildQueryString(null, CodecRegistry.DEFAULT_INSTANCE);
-            return maybeAddSemicolon(queryString).toString();
+            // try first with all values inlined (will not work if some values require custom codecs)
```

Comment:
Sorry, I should have explained better my rationale:

1. First, try to generate a query string avoiding bind markers as much as possible (because `UPDATE t SET x = 'foo' WHERE y = 42` is more precise than `UPDATE t SET x = ? WHERE y = ?`);
2. If some variable causes a `CodecNotFoundException` to be raised, try again without the values (because `UPDATE t SET x = ? WHERE y = ?` is still better than an error message);
3. If that raises an error again, then return the dummy `built query ...` error message.

But you are right that if `forceNoValues = true`, there is no reason why we should not use the cache. I will amend that.